### PR TITLE
feat/node: More type hints

### LINKED ISSUE #796: Assign numeric strings a type instead of `string` in node.js
## Description

Some values, e.g. the amount of base tokens, are serialized as a decimal numeric string. Using a type, for example: 
```typescript 
declare type NumericString = string;
```
 to annotate them would improve readability.

Probably, this change will affect the external API in a non-breaking way.

## Motivation

Confusion between hex encoded strings and numeric strings.

## Are you planning to do it yourself in a pull request?
Maybe

### PR DESCRIPTION
**Pull request by kwek20**
# Description of change

Updates code to use a lot more type hints to node.
Added a NumericString specific for numbers as a string

## Links to any relevant issues

fixes #796 

## Type of change

- Bug fix (a non-breaking change which fixes an issue)
- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

No actual changes made as all type aliases refer to the correct type internally (string)
Ran tests and a few examples to make sure.

### CODE REVIEW

File: bindings/nodejs/lib/client/client.ts [hunk 1]
```diff
@@ -700,7 +707,10 @@ export class Client {
      * @param bech32Hrp The Bech32 HRP (human readable part) to be used.
      * @returns The corresponding Bech32 address.
      */
-    async hexToBech32(hex: string, bech32Hrp?: string): Promise<string> {
+    async hexToBech32(
+        hex: string,
```
Comment (DaughterOfMars) [inline bindings/nodejs/lib/client/client.ts]: HexEncodedString?

File: bindings/nodejs/lib/client/client.ts [hunk 2]
```diff
@@ -700,7 +707,10 @@ export class Client {
...
+        bech32Hrp?: string,
+    ): Promise<HexEncodedString> {
```
Comment (DaughterOfMars) [inline bindings/nodejs/lib/client/client.ts]: This shouldn't be a HexEncodedString

File: bindings/nodejs/lib/client/client.ts [hunk 3]
```diff
@@ -741,7 +751,10 @@ export class Client {
      * @param bech32Hrp The Bech32 HRP (human readable part) to be used.
      * @returns The corresponding Bech32 address.
      */
-    async nftIdToBech32(nftId: string, bech32Hrp?: string): Promise<string> {
+    async nftIdToBech32(
+        nftId: NftId,
+        bech32Hrp?: string,
+    ): Promise<HexEncodedString> {
```
Comment (DaughterOfMars) [inline bindings/nodejs/lib/client/client.ts]: Not a HexEncodedString

File: bindings/nodejs/lib/client/client.ts [hunk 4]
```diff
@@ -763,7 +776,7 @@ export class Client {
     async hexPublicKeyToBech32Address(
         hex: string,
         bech32Hrp?: string,
-    ): Promise<string> {
+    ): Promise<HexEncodedString> {
```
Comment (DaughterOfMars) [inline bindings/nodejs/lib/client/client.ts]: Same

File: bindings/nodejs/lib/client/client.ts [hunk 5]
```diff
@@ -1211,7 +1224,7 @@ export class Client {
      */
     async requestFundsFromFaucet(
         url: string,
-        address: string,
+        address: HexEncodedString,
```
Comment (DaughterOfMars) [inline bindings/nodejs/lib/client/client.ts]: Is this a bech32 address? If so, not Hex encoded

File: bindings/nodejs/lib/types/models/api/plugins/indexer/output-response.ts [hunk 6]
```diff
@@ -13,11 +14,11 @@ export interface IOutputsResponse {
     /**
      * The maximum count of results that are returned by the node.
      */
-    pageSize: string;
+    pageSize: NumericString;
     /**
      * The cursor to use for getting the next results.
      */
-    cursor?: string;
+    cursor?: NumericString;
```
Comment (DaughterOfMars) [inline bindings/nodejs/lib/types/models/api/plugins/indexer/output-response.ts]: Not sure cursors are numeric strings

**Review by DaughterOfMars — COMMENTED** [review]
For IDs we should probably distinguish between hex and prefixed hex strings

File: bindings/nodejs/lib/types/models/api/plugins/indexer/output-response.ts (hunk 6 above)
Comment (kwek20) [reply bindings/nodejs/lib/types/models/api/plugins/indexer/output-response.ts]:
```
    /// Starts the search from the cursor (confirmationMS+outputId.pageSize).
    Cursor(String),
   ```
   Seems to be

Comment (kwek20) [reply bindings/nodejs/lib/types/models/api/plugins/indexer/output-response.ts]: never mind, youre just better!! Fixed 👌

File: bindings/nodejs/lib/types/block/address.ts [hunk 7]
```diff
@@ -5,6 +5,11 @@ import { plainToInstance } from 'class-transformer';
 import { HexEncodedString } from '../utils';
 import { AliasId, NftId } from './id';
 
+/**
+ * A hex encoded address prepended by its network type.
+ */
+type Bech32 = string;
```
Comment (DaughterOfMars) [inline bindings/nodejs/lib/types/block/address.ts]: `Bech32Address`

File: bindings/nodejs/lib/types/client/output_builder_params/alias-output-params.ts [hunk 8]
```diff
@@ -1,6 +1,6 @@
 // Copyright 2021-2023 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
-import { Feature, HexEncodedString } from '../..';
+import { AliasId, Feature, HexEncodedString } from '../..';
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/client/output_builder_params/alias-output-params.ts:4]:
```suggestion

import { AliasId, Feature, HexEncodedString } from '../..';
```

File: bindings/nodejs/lib/types/client/output_builder_params/basic-output-params.ts [hunk 9]
```diff
@@ -1,6 +1,6 @@
 // Copyright 2021-2023 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
-import { UnlockCondition, Feature, INativeToken } from '../..';
+import { UnlockCondition, Feature, INativeToken, NumericString } from '../..';
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/client/output_builder_params/basic-output-params.ts:4]:
```suggestion

import { UnlockCondition, Feature, INativeToken, NumericString } from '../..';
```

File: bindings/nodejs/lib/types/client/output_builder_params/nft-output-params.ts [hunk 10]
```diff
@@ -1,6 +1,6 @@
 // Copyright 2021-2023 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
-import { Feature, HexEncodedString } from '../..';
+import { Feature, NftId } from '../..';
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/client/output_builder_params/nft-output-params.ts:4]:
```suggestion

import { Feature, NftId } from '../..';
```

File: bindings/nodejs/lib/types/block/output/output.ts [hunk 11]
```diff
@@ -35,15 +35,15 @@ enum OutputType {
  * The base class for outputs.
  */
 abstract class Output /*implements ICommonOutput*/ {
-    readonly amount: string;
+    readonly amount: NumericString;
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/block/output/output.ts:38]: Don't we use `u64` for things like this ?

File: bindings/nodejs/lib/types/models/api/plugins/indexer/output-response.ts [hunk 12]
```diff
@@ -1,6 +1,7 @@
 // Copyright 2023 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
 
+import { NumericString } from '../../../../utils';
 import type { HexEncodedString } from '../../../../utils/hex-encoding';
 /**
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/models/api/plugins/indexer/output-response.ts:7]:
```suggestion

/**
```

File: bindings/nodejs/lib/types/models/info/node-info-protocol.ts [hunk 13]
```diff
@@ -1,6 +1,7 @@
 // Copyright 2023 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
 
+import { NumericString } from '../../utils/numeric';
 import type { IRent } from '../rent';
 /**
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/models/info/node-info-protocol.ts:7]:
```suggestion

/**
```

File: bindings/nodejs/lib/types/block/output/output.ts (hunk 11 above)
Comment (kwek20) [reply bindings/nodejs/lib/types/block/output/output.ts:38]: no its not bigint yet, thats 2.0

**Review by DaughterOfMars — DISMISSED** [review]
very nice

File: bindings/nodejs/lib/wallet/account.ts [hunk 14]
```diff
@@ -1190,8 +1203,8 @@ export class Account {
      * @returns The sent transaction.
      */
     async send(
-        amount: bigint | string,
-        address: string,
+        amount: bigint | NumericString,
+        address: HexEncodedString,
```
Comment (Thoralf-M) [inline bindings/nodejs/lib/wallet/account.ts]:
```suggestion
        address: Bech32Address,
```
Probably also needs an import

File: bindings/nodejs/lib/types/block/address.ts [hunk 15]
```diff
@@ -5,6 +5,11 @@ import { plainToInstance } from 'class-transformer';
 import { HexEncodedString } from '../utils';
 import { AliasId, NftId } from './id';
 
+/**
+ * A hex encoded address prepended by its network type.
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/block/address.ts]: It's not really hex encoded

File: bindings/nodejs/lib/wallet/account.ts [hunk 16]
```diff
@@ -47,9 +47,21 @@ import {
...
+    FoundryId,
+    TransactionId,
+    NumericString,
+    Bech32Address,
 } from '../types';
 import { plainToInstance } from 'class-transformer';
-import { bigIntToHex, hexToBigInt } from '../types/utils/hex-encoding';
+import {
+    bigIntToHex,
+    HexEncodedString,
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/wallet/account.ts]: Defined but not used

File: bindings/nodejs/lib/utils/utils.ts [hunk 17]
```diff
@@ -288,7 +292,7 @@ export class Utils {
      * @param bech32Hrp The Bech32 HRP (human readable part) to use.
      * @returns The Bech32-encoded address string.
      */
-    static nftIdToBech32(nftId: string, bech32Hrp: string): string {
+    static nftIdToBech32(nftId: string, bech32Hrp: string): Bech32Address {
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/utils/utils.ts]:
```suggestion
    static nftIdToBech32(nftId: NftId, bech32Hrp: string): Bech32Address {
```

**Review by thibault-martinez — COMMENTED** [review]
Even if not fix/feature, I would put a little entry in the changelog

**Review by thibault-martinez — APPROVED** [review]

**Review by Thoralf-M — APPROVED** [review]

**Pull request merged**